packages/react-router/lib/dom/ssr/routeModules.ts

```diff
@@ -62,26 +125,30 @@ export type ClientLoaderFunctionArgs = LoaderFunctionArgs<undefined> & {
 };
 
 /**
- * ErrorBoundary to display for this route
+ * Parameters passed to the [`headers`]{@link HeadersFunction} function
  */
-export type ErrorBoundaryComponent = ComponentType;
+export type HeadersArgs = {
+  loaderHeaders: Headers;
+  parentHeaders: Headers;
+  actionHeaders: Headers;
+  errorHeaders: Headers | undefined;
+};
 
 /**
- * `<Route HydrateFallback>` component to render on initial loads
- * when client loaders are present
+ * A function that returns HTTP headers to be used for a route. These headers
+ * will be merged with (and take precedence over) headers from parent routes.
  */
-export type HydrateFallbackComponent = ComponentType;
+export interface HeadersFunction {
+  (args: HeadersArgs): Headers | HeadersInit;
+}
 
 /**
  * Optional, root-only `<Route Layout>` component to wrap the root content in.
  * Useful for defining the <html>/<head>/<body> document shell shared by the
  * Component, HydrateFallback, and ErrorBoundary
  */
-export type LayoutComponent = ComponentType<{
-  children: ReactElement<
-    unknown,
-    ErrorBoundaryComponent | HydrateFallbackComponent | RouteComponent
-  >;
+export type LayoutComponent = React.ComponentType<{
```

Comment:
I legit didn't know this type existed, FWIW I added it here and got a comment from Michael to maybe change it to a different React type, check it out here:
https://github.com/remix-run/react-router/pull/12543#discussion_r1888106592